### PR TITLE
pam_motd: filter motd by user and group

### PR DESCRIPTION
**Pull request by ikerexxe**
modules/pam_motd/pam_motd.c: filter motd by user and group owning the
proper files. This is achieved by changing the ids of the process
reading the files from root to the target user.

Resolves:
https://bugzilla.redhat.com/show_bug.cgi?id=1861640

### CODE REVIEW

File: modules/pam_motd/pam_motd.c [hunk 1]
```diff
@@ -282,6 +283,97 @@ static void try_to_display_directories_with_overrides(pam_handle_t *pamh,
...
+    if (setuid(pw->pw_uid) < 0) {
+        pam_syslog(pamh, LOG_ERR, "Unable to change UID to %d\n", pw->pw_uid);
+
+        retval = PAM_SESSION_ERR;
+        goto out;
+    }
+
+out:
+    return retval;
+}
```
Comment (ldv-alt) [inline modules/pam_motd/pam_motd.c:316]: Please consider using pam_modutil_drop_priv/pam_modutil_regain_priv interface.

**Review by ldv-alt — CHANGES_REQUESTED** [review]

File: modules/pam_motd/pam_motd.c (hunk 1 above)
Comment (ikerexxe) [reply modules/pam_motd/pam_motd.c]: Thanks! Now the code is neater.

**Comment by ikerexxe** [comment]
@ldv-alt do you have any feedback now that pam_modutil_*_priv is being used?

**Comment by ldv-alt** [comment]
On Wed, Sep 23, 2020 at 01:28:20AM -0700, Iker Pedrosa wrote:
> @ldv-alt do you have any feedback now that `pam_modutil_*_priv` is being used?

Sorry, I was distracted by other things.

I have three questions/concerns:

- If motd_path_split == NULL, pam_modutil_regain_priv() will be invoked
without a prior pam_modutil_drop_priv() call; I don't think
pam_modutil_regain_priv() would be happy about it.

- If motd_path_split == NULL and motd_dir_path_split != NULL, then
try_to_display_directories_with_overrides() will be invoked without a prior
drop_privileges() call; is this intended?

- In a very unlikely case when pam_modutil_regain_priv returns an error,
the process will be left in an inconsistent state; the module should fail
with an error in that case, shouldn't it?

File: modules/pam_motd/pam_motd.c [hunk 2]
```diff
@@ -358,35 +424,23 @@ int pam_sm_open_session(pam_handle_t *pamh, int flags,
...
     _pam_drop(motd_dir_path_split);
 
-    retval = pam_putenv(pamh, "MOTD_SHOWN=pam");
-
-    return retval == PAM_SUCCESS ? PAM_IGNORE : retval;
+    if (retval == PAM_SUCCESS) {
+        retval = pam_putenv(pamh, "MOTD_SHOWN=pam");
+        return retval == PAM_SUCCESS ? PAM_IGNORE : retval;
+    } else {
+        pam_putenv(pamh, "MOTD_SHOWN=pam");
```
Comment (ikerexxe) [inline modules/pam_motd/pam_motd.c:441]: I'm not sure if this should be set in case of failure. What's your opinion?

**Comment by ikerexxe** [comment]
> On Wed, Sep 23, 2020 at 01:28:20AM -0700, Iker Pedrosa wrote: @ldv-alt do you have any feedback now that `pam_modutil_*_priv` is being used?
> Sorry, I was distracted by other things. I have three questions/concerns: - If motd_path_split == NULL, pam_modutil_regain_priv() will be invoked without a prior pam_modutil_drop_priv() call; I don't think pam_modutil_regain_priv() would be happy about it. - If motd_path_split == NULL and motd_dir_path_split != NULL, then try_to_display_directories_with_overrides() will be invoked without a prior drop_privileges() call; is this intended? - In a very unlikely case when pam_modutil_regain_priv returns an error, the process will be left in an inconsistent state; the module should fail with an error in that case, shouldn't it?

I agree with all of your points. Thank you for the feedback.

**Comment by ldv-alt** [comment]
On Thu, Sep 24, 2020 at 01:53:12AM -0700, Iker Pedrosa wrote:
> @ikerexxe commented on this pull request.
> >  
>    out:
>      _pam_drop(motd_path_copy);
>      _pam_drop(motd_path_split);
>      _pam_drop(motd_dir_path_copy);
>      _pam_drop(motd_dir_path_split);
>  
> -    retval = pam_putenv(pamh, "MOTD_SHOWN=pam");
> -
> -    return retval == PAM_SUCCESS ? PAM_IGNORE : retval;
> +    if (retval == PAM_SUCCESS) {
> +        retval = pam_putenv(pamh, "MOTD_SHOWN=pam");
> +        return retval == PAM_SUCCESS ? PAM_IGNORE : retval;
> +    } else {
> +        pam_putenv(pamh, "MOTD_SHOWN=pam");
> 
> I'm not sure if this should be set in case of failure. What's your opinion?

I'd rather be on the safe side and wouldn't set MOTD_SHOWN if the module
is going to return an error.

**Comment by ikerexxe** [comment]
> On Thu, Sep 24, 2020 at 01:53:12AM -0700, Iker Pedrosa wrote: @ikerexxe commented on this pull request. > out: _pam_drop(motd_path_copy); _pam_drop(motd_path_split); _pam_drop(motd_dir_path_copy); _pam_drop(motd_dir_path_split); - retval = pam_putenv(pamh, "MOTD_SHOWN=pam"); - - return retval == PAM_SUCCESS ? PAM_IGNORE : retval; + if (retval == PAM_SUCCESS) { + retval = pam_putenv(pamh, "MOTD_SHOWN=pam"); + return retval == PAM_SUCCESS ? PAM_IGNORE : retval; + } else { + pam_putenv(pamh, "MOTD_SHOWN=pam"); I'm not sure if this should be set in case of failure. What's your opinion?
> I'd rather be on the safe side and wouldn't set MOTD_SHOWN if the module is going to return an error.

Done!

File: modules/pam_motd/pam_motd.c [hunk 3]
```diff
@@ -282,6 +282,72 @@ static void try_to_display_directories_with_overrides(pam_handle_t *pamh,
...
+    const char *username;
+    int retval;
+
+    retval = pam_get_user(pamh, &username, "key user");
+
+    if (retval == PAM_SUCCESS) {
+        pw = pam_modutil_getpwnam (pamh, username);
+    } else {
+        return PAM_SESSION_ERR;
+    }
```
Comment (ldv-alt) [inline modules/pam_motd/pam_motd.c:297]: Please check the value returned by pam_modutil_getpwnam(); if it's NULL, then pam_modutil_drop_priv() shouldn't be called.

**Review by ldv-alt — CHANGES_REQUESTED** [review]

File: modules/pam_motd/pam_motd.c (hunk 3 above)
Comment (ikerexxe) [reply modules/pam_motd/pam_motd.c:297]: Done

**Review by ldv-alt — APPROVED** [review]

**Comment by ldv-alt** [comment]
I think it's OK now, thanks.

**Comment by ikerexxe** [comment]
Do you think this is good enough to merge or do you want another pair of eyes to review it?

**Comment by t8m** [comment]
The code is good. However I strongly believe this change should be reflected in pam_motd manual page and in the NEWS file. So please add that documentation as a separate patch and I'll merge this.

**Comment by ikerexxe** [comment]
I've created a new release section in the NEWS file, but I'm not sure if it should be 1.5.0 (as written) or 1.4.1

File: modules/pam_motd/pam_motd.8.xml [hunk 4]
```diff
@@ -64,8 +64,9 @@
       override files with the same name in <filename>/usr/lib/motd.d/</filename>.
     </para>
     <para>
-      Files the in the directories listed above are displayed in
-      lexicographic order by name.
+      Files in the directories listed above are displayed in lexicographic
+      order by name. Moreover, the files are filtered by checking the ownership
```
Comment (t8m) [inline modules/pam_motd/pam_motd.8.xml:68]: I'd use something like: "files are filtered by reading them with the credentials of the target user authenticating on the system."

File: NEWS [hunk 5]
```diff
@@ -1,5 +1,8 @@
 Linux-PAM NEWS -- history of user-visible changes.
 
+Release 1.5.0
+* pam_motd: filter motd by user and group owning the files.
```
Comment (t8m) [inline NEWS:4]: I prefer something like: "Read motd files with target user credentials skipping unreadable ones."

**Comment by t8m** [comment]
> I've created a new release section in the NEWS file, but I'm not sure if it should be 1.5.0 (as written) or 1.4.1

That does not matter and can be changed later. But yeah 1.5.0 is more likely.

**Review by t8m — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged, thank you for the contribution.

File: modules/pam_motd/pam_motd.c [hunk 6]
```diff
@@ -282,6 +282,72 @@ static void try_to_display_directories_with_overrides(pam_handle_t *pamh,
     _pam_drop(dirscans);
 }
 
+static int drop_privileges(pam_handle_t *pamh, struct pam_modutil_privs *privs)
+{
+    struct passwd *pw;
+    const char *username;
+    int retval;
+
+    retval = pam_get_user(pamh, &username, "key user");
```
Comment (ldv-alt) [inline modules/pam_motd/pam_motd.c:291]: Apparently, I missed this "key user" argument.
I suppose NULL should be passed to pam_get_user instead.

Comment (t8m) [reply modules/pam_motd/pam_motd.c:291]: Ah, good catch. Will you submit a patch?

Comment (ikerexxe) [reply modules/pam_motd/pam_motd.c:291]: Thank you! I think I didn't use the correct example to drop/regain privileges.

Comment (ikerexxe) [reply modules/pam_motd/pam_motd.c:291]: Created PR: https://github.com/linux-pam/linux-pam/pull/282

**Pull request merged**